File: Week 1 C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// CS50 problem set 1 - Credit
int main(void)
{
    long input = get_long("Number: ");
    // Duplicate input for counting loop
    long num = input;
    // Initialize count
    int count = 0;
    // Count Digits
    while (num != 0)
    {
        num = num / 10;
        count++;
    }
    // Check to ensure correct number digits
    if (count != 13 && count != 15 && count != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    // Initialize variables
    int s1 = 0, s2 = 0, d1, d2, m1, m2;
    long i = input;
    // Perform Checksum
    do
    {
        // Calculate sum of first digits
        d1 = i % 10;
        s1 = s1 + d1;
        // Move to next digit
        i = i / 10;
        // Calculate sum of second digits
        d2 = i % 10;
        d2 = d2 * 2;
        // Manipulate second digits as to sum digits if product > 9
        m1 = d2 % 10;
        m2 = d2 / 10;
        s2 = s2 + m1 + m2;
        // Move to next digit for loop
        i = i / 10;
    }
    while (i != 0);
    int total = s1 + s2;
    // Check Luhn Algorithm
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    // If amex card, strip 13 characters and check number
    // Initialize card to declare invalid if all card types fail
    int card = 0;
    int amex = input / 10000000000000;
    if (amex == 34 || amex == 37)
    {
        printf("AMEX\n");
        card = 1;
    }
    // If mastercard, strip 14 characters and check number
    int mc = input / 100000000000000;
    if (mc > 50 && mc < 56)
    {
        printf("MASTERCARD\n");
        card = 1;
    }
    // If visa, strip 12 and 15 characters and check number for each
    int visa1 = input / 1000000000000;
    int visa2 = input / 1000000000000000;
    if (visa1 == 4 || visa2 == 4)
    {
        printf("VISA\n");
        card = 1;
    }
    // State invalid if not matching card found
    if (card == 0)
    {
        printf("INVALID\n");
    }
}

File: Week 6 Python/sentimental-credit/credit.py
from cs50 import get_int
import re

# Get card number
cc_int = get_int('Number: ')
# Convert number to string for manipulation
cc = str(cc_int)

# Initialize variables
sum1 = 0
sum2 = 0
# Loop through card number and use Luhn's Algorithm to determine validity
for i in range(len(cc)):
    # Get last digit of card
    y = cc_int % 10
    # Remove last digit of card
    cc_int = cc_int // 10
    # Get original second from last digit of card and multiply by two
    x = 2 * (cc_int % 10)
    # If x is multi-digit add each separately and keep running total
    sum1 += (x % 10) + (x // 10)
    # Keep running total of sums of y
    sum2 += y
    # Remove original second from last digit of card for next iteration
    cc_int = cc_int // 10

# Get total of both sums
sum_tot = sum1 + sum2
# Determine if card is a valid amex, visa, or mastercard
# Test Luhn's Algorithm to see if last digit of total sum is 0
if sum_tot % 10 == 0:
    # Test first digits and length of card to determine card type or print invalid if no match found
    if (re.search('\A34', cc) or re.search('\A37', cc)) and len(cc) == 15:
        print('AMEX')
    elif cc[0] == '5' and int(cc[1]) > 0 and int(cc[1]) < 6 and len(cc) == 16:
        print('MASTERCARD')
    elif cc[0] == '4' and (len(cc) == 13 or len(cc) == 16):
        print('VISA')
    else:
        print('INVALID')
# Invalid if no match found or fails Luhn's Algorithm
else:
    print('INVALID')